File: zorpctl/szig.py
############################################################################
##
## Copyright (c) 2000-2014 BalaBit IT Ltd, Budapest, Hungary
##
## This program is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License
## as published by the Free Software Foundation; either version 2
## of the License, or (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program; if not, write to the Free Software
## Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
##
############################################################################

import socket
from zorpctl.SZIGMessages import (
        MessageAuthorizeAccept, MessageAuthorizeReject,
        MessageGetChild, MessageGetDeadLockCheck,
        MessageGetLogLevel, MessageGetLogSpec,
        MessageGetSibling, MessageGetValue,
        MessageReload, MessageReloadResult,
        MessageSetDeadLockCheck, MessageSetLogLevel,
        MessageSetLogSpec, MessageStopSession
    )
from zorpctl.ZorpctlConf import ZorpctlConfig


class Response(object):
    def __init__(self, succeeded, value = None):
        self.is_succeeded = succeeded
        self.value = value

class ResponseDeadlockCheck(Response):

    def isSet(self):
        return self.value == "1"

class Handler(object):
    """
    Class created for handling messages sent by Szig to Zorp
    and receiving answer from Zorp
    """

    _success_prefix = "OK "
    _fail_prefix = "FAIL "

    def __init__(self, server_address):
        self.max_command_length = 4096
        self.response_length = 4096
        self.socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        self.server_address = server_address
        try:
            self.socket.connect(self.server_address)
        except IOError as e:
            e.message = "Socket not found, %s" % server_address
            raise e

    def talk(self, message):
        """
        Sends an instance of Message and
        returns the response as Response class
        """
        self.send(message)
        return self.recv()

    def send(self, message):
        """
        Sending a message to Zorp.
        Messages can be derived from abstract Message class.
        """
        self._write_request(str(message))

    def recv(self):
        """
        Returns an instance of Response class.
        """
        resp = self._read_response()
        return Response(self._isSucceeded(resp), self._cutPrefix(resp))

    def _write_request(self, request):
        """
        Writing a command message to a Unix Domain Socket
        to communicate with Zorp.

        Raises SZIGError if not all the data has been sent.
        SZIGError value is a tuple of sent/all
        """

        request_length = len(request)
        if request_length > self.max_command_length:
            raise SZIGError("Given request is longer than %s" % self.max_command_length)

        sent_data_length = self.socket.send(request)
        if sent_data_length < request_length:
            msg = "There was an error while sending the request (%s/%s)!" % (sent_data_length, request_length)
            raise SZIGError(msg, (sent_data_length, request_length))

    def _read_response(self, resp_len = None):
        """
        Reading from a Unix Domain Socket
        to communicate with Zorp.
        """
        if not resp_len:
            resp_len = self.response_length
        if resp_len < 1:
            raise SZIGError("Response length should be greater than 0")

        response = self.socket.recv(resp_len)
        if not response:
            raise SZIGError("There was an error while receiving the answer!")

        return response[:-1] if response[-1:] == '\n' else response

    def _isSucceeded(self, response):
        """
        Method for checking if Zorp understood
        the given request by inspecting the response.
        """
        return response[:len(self._success_prefix)] == self._success_prefix

    def _cutPrefix(self, string):
        """
        Cuts the defined prefix from a string.
        """
        if string[:len(self._success_prefix)] == self._success_prefix:
            string =  string[len(self._success_prefix):]
        else:
            if string[:len(self._fail_prefix)] == self._fail_prefix:
                string = string[len(self._fail_prefix):]

        return string

class SZIG(object):

    def __init__(self, process_name, handler=None):
        ZORPCTLCONF = ZorpctlConfig.Instance()
        self.pidfile_dir = ZORPCTLCONF['ZORP_PIDFILEDIR']
        if not handler:
            handler = Handler
        self.handler = handler(self.pidfile_dir + '/zorpctl.' + process_name)

    def get_value(self, key):
        response = self.handler.talk(MessageGetValue(key))
        return None if response.value == "None" else response.value

    def get_sibling(self, node):
        response =  self.handler.talk(MessageGetSibling(node))
        return None if response.value == "None" else response.value

    def get_child(self, node):
        response =  self.handler.talk(MessageGetChild(node))
        return None if response.value == "None" else response.value

    @property
    def loglevel(self):
        self.handler.send(MessageGetLogLevel())
        return int(self.handler.recv().value)

    @loglevel.setter
    def loglevel(self, value):
        self.handler.send(MessageSetLogLevel(value))
        if not self.handler.recv().is_succeeded:
            raise SZIGError("Log level has not been set.")

    @property
    def logspec(self):
        self.handler.send(MessageGetLogSpec())
        return self.handler.recv().value

    @logspec.setter
    def logspec(self, value):
        """
        Setting LOGSPEC expecting a log specification
        string as value
        """
        self.handler.send(MessageSetLogSpec(value))
        if not self.handler.recv().is_succeeded:
            raise SZIGError("Log specification has not been set.")

    @property
    def deadlockcheck(self):
        self.handler.send(MessageGetDeadLockCheck())
        dlc = self.handler.recv()
        dlc.__class__ = ResponseDeadlockCheck

        return dlc.isSet()

    @deadlockcheck.setter
    def deadlockcheck(self, value):
        """
        Sets Deadlock Check, expects a boolean as value.
        """
        self.handler.talk(MessageSetDeadLockCheck(value))

    def reload(self):
        self.handler.talk(MessageReload())

    def reload_result(self):
        result = self.handler.talk(MessageReloadResult())
        return result.is_succeeded

    def stop_session(self, instance):
        response = self.handler.talk(MessageStopSession(instance))
        if not response.is_succeeded:
            raise SZIGError("Session stop failed! Response was: %s" % response.value)

    def authorize_accept(self, session_id, description):
        response = self.handler.talk(MessageAuthorizeAccept(session_id, description))
        if not response.is_succeeded:
            raise SZIGError(response.value)
        return response.value

    def authorize_reject(self, session_id, description):
        response = self.handler.talk(MessageAuthorizeReject(session_id, description))
        if not response.is_succeeded:
            raise SZIGError(response.value)
        return response.value

class SZIGError(Exception):
    """
    Exception Class created for Szig specific errors.
    """
    def __init__(self, msg, value = None):
        self.msg = msg
        self.value = value

    def __str__(self):
        return self.msg + repr(self.value)

File: zorpctl_in_cc/szig.cc
/***************************************************************************
 *
 * Copyright (c) 2000-2015 BalaBit IT Ltd, Budapest, Hungary
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 ***************************************************************************/

#include "szig.h"

#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>
#include <stdio.h>
#include <glib.h>

static const int SZIG_MAX_CMD_LENGTH = 131072 + 256; /* 128KiB is the max length of a single argument + a little room */

static int
z_szig_write_request(ZSzigContext *ctx, const char *request)
{
  if (write(ctx->fd, request, strlen(request)) < 0)
    return 0;
  return 1;
}

static int
z_szig_read_response(ZSzigContext *ctx, char *response, size_t response_len)
{
  int len;
  if ((len = read(ctx->fd, response, response_len - 1)) < 0)
    return 0;
  response[len] = 0;
  if (response[len - 1] == '\n')
    response[len-1] = 0;
  return 1;
}

int
z_szig_get_value(ZSzigContext *ctx, const char *key, char *result, size_t result_len)
{
  char buf[SZIG_MAX_CMD_LENGTH];

  snprintf(buf, sizeof(buf), "GETVALUE %s\n", key);
  if (!z_szig_write_request(ctx, buf))
    return 0;
  if (!z_szig_read_response(ctx, result, result_len))
    return 0;
  return 1;
}

int
z_szig_get_sibling(ZSzigContext *ctx, const char *key, char *result, size_t result_len)
{
  char buf[SZIG_MAX_CMD_LENGTH];

  snprintf(buf, sizeof(buf), "GETSBLNG %s\n", key);
  if (!z_szig_write_request(ctx, buf))
    return 0;
  if (!z_szig_read_response(ctx, result, result_len))
    return 0;

  return 1;
}

int
z_szig_get_child(ZSzigContext *ctx, const char *key, char *result, size_t result_len)
{
  char buf[SZIG_MAX_CMD_LENGTH];

  snprintf(buf, sizeof(buf), "GETCHILD %s\n", key);
  if (!z_szig_write_request(ctx, buf))
    return 0;
  if (!z_szig_read_response(ctx, result, result_len))
    return 0;

  return 1;
}

int
z_szig_logging(ZSzigContext *ctx, const char *subcmd, const char *param, char *result, size_t result_len)
{
  char buf[SZIG_MAX_CMD_LENGTH];
  char res_buf[128];

  snprintf(buf, sizeof(buf), "LOGGING %s %s\n", subcmd, param);
  if (!z_szig_write_request(ctx, buf))
    return 0;
  if (!z_szig_read_response(ctx, res_buf, sizeof(res_buf)))
    return 0;

  if (strncmp(res_buf, "FAIL ", 5) == 0)
    {
      return 0;
    }
  else if (strncmp(res_buf, "OK ", 3) == 0)
    {
      if (result)
        {
          strncpy(result, res_buf + 3, result_len);
          res_buf[result_len - 1] = 0;
        }
      return 1;
    }
  else
    {
      return 0;
    }
}

int
z_szig_deadlockcheck(ZSzigContext *ctx, const char *subcmd, char *result, size_t result_len)
{
  char buf[SZIG_MAX_CMD_LENGTH];
  char res_buf[128];

  snprintf(buf, sizeof(buf), "DEADLOCKCHECK %s\n", subcmd);
  if (!z_szig_write_request(ctx, buf))
    return 0;
  if (!z_szig_read_response(ctx, res_buf, sizeof(res_buf)))
    return 0;

  if (strncmp(res_buf, "FAIL ", 5) == 0)
    {
      return 0;
    }
  else if (strncmp(res_buf, "OK ", 3) == 0)
    {
      if (result)
        {
          strncpy(result, res_buf + 3, result_len);
          res_buf[result_len - 1] = 0;
        }
      return 1;
    }
  else
    {
      return 0;
    }
}

int
z_szig_reload(ZSzigContext *ctx, const char *subcmd, char *result, size_t result_len)
{
  char buf[SZIG_MAX_CMD_LENGTH];
  char res_buf[128];

  if (!subcmd)
    snprintf(buf, sizeof(buf), "RELOAD\n");
  else
    snprintf(buf, sizeof(buf), "RELOAD %s\n", subcmd);

  if (!z_szig_write_request(ctx, buf))
    return 0;
  if (!z_szig_read_response(ctx, res_buf, sizeof(res_buf)))
    return 0;

  if (strncmp(res_buf, "FAIL ", 5) == 0)
    {
      return 0;
    }
  else if (strncmp(res_buf, "OK ", 3) == 0)
    {
      if (result)
        {
          strncpy(result, res_buf + 3, result_len);
          res_buf[result_len - 1] = 0;
        }
      return 1;
    }
  else
    {
      return 0;
    }
}


int
z_szig_authorize(ZSzigContext *ctx, const char *instance, int accept, const char *description, char *result, size_t result_len)
{
  char buf[SZIG_MAX_CMD_LENGTH];
  char res_buf[128];

  if (!instance)
    return 0;
  else
    snprintf(buf, sizeof(buf), "AUTHORIZE %s %s %s\n", accept ? "ACCEPT" : "REJECT", instance, description);

  if (!z_szig_write_request(ctx, buf))
    return 0;
  if (!z_szig_read_response(ctx, res_buf, sizeof(res_buf)))
    return 0;

  if (strncmp(res_buf, "FAIL ", 5) == 0)
    {
      if (result)
        {
          strncpy(result, res_buf + 5, result_len);
          res_buf[result_len - 1] = 0;
        }
      return 0;
    }
  else if (strncmp(res_buf, "OK ", 3) == 0)
    {
      if (result)
        {
          strncpy(result, res_buf + 3, result_len);
          res_buf[result_len - 1] = 0;
        }
      return 1;
    }
  else
    {
      return 0;
    }
}

int
z_szig_coredump(ZSzigContext *ctx)
{
  char buf[SZIG_MAX_CMD_LENGTH];
  char res_buf[128];

  snprintf(buf, sizeof(buf), "COREDUMP\n");
  if (!z_szig_write_request(ctx, buf))
    return 0;
  if (!z_szig_read_response(ctx, res_buf, sizeof(res_buf)))
    return 0;

  if (strncmp(res_buf, "FAIL ", 5) == 0)
    {
      return 0;
    }
  else if (strncmp(res_buf, "OK ", 3) == 0)
    {
      return 1;
    }
  else
    {
      return 0;
    }
}

ZSzigContext *
z_szig_context_new(const char *instance_name)
{
  ZSzigContext *ctx;
  struct sockaddr_un unaddr;
  int fd;

  fd = socket(AF_UNIX, SOCK_STREAM, 0);
  if (fd == -1)
    return NULL;
  unaddr.sun_family = AF_UNIX;
  snprintf(unaddr.sun_path, sizeof(unaddr.sun_path), ZORP_PIDFILEDIR "zorpctl.%s", instance_name);
  if (connect(fd, (struct sockaddr *) &unaddr, sizeof(unaddr)) < 0)
    {
      close(fd);
      return NULL;
    }

  ctx = g_new(ZSzigContext, 1);
  ctx->fd = fd;
  return ctx;
}

void
z_szig_context_destroy(ZSzigContext *ctx)
{
  if (ctx)
    {
      close(ctx->fd);
      g_free(ctx);
    }
}
